library/kani_macros/src/sysroot/loop_contracts/mod.rs

```diff
@@ -0,0 +1,101 @@
+//!
+
+use proc_macro::TokenStream;
+use proc_macro_error2::abort_call_site;
+use quote::{format_ident, quote};
+use syn::spanned::Spanned;
+use syn::token::AndAnd;
+use syn::{BinOp, Expr, ExprBinary, Stmt};
+
+/// Expand loop contracts macros.
+///
+/// A while loop of the form
+/// ``` rust
+///  while guard {
+///      body
+///  }
+/// ```
+/// will be annotated as
+/// ``` rust
+/// #[inline(never)]
+/// #[kanitool::fn_marker = "kani_register_loop_contract"]
+/// const fn kani_register_loop_contract_id<T, F: FnOnce() -> T>(f: F) -> T {
+///     unreachable!()
+/// }
+///  while kani_register_loop_contract_id(|| -> bool {inv};) && guard {
+///      body
+///  }
+/// ```
+pub fn loop_invariant(attr: TokenStream, item: TokenStream) -> TokenStream {
+    // parse the stmt of the loop
+    let mut loop_stmt: Stmt = syn::parse(item.clone()).unwrap();
+
+    // name of the loop invariant as closure of the form
+    // __kani_loop_invariant_#startline_#startcol_#endline_#endcol
+    let mut inv_name: String = "__kani_loop_invariant".to_owned();
+    let loop_id = generate_unique_id_from_span(&loop_stmt);
+    inv_name.push_str(&loop_id);
+
+    // expr of the loop invariant
+    let inv_expr: Expr = syn::parse(attr).unwrap();
+
+    // ident of the register function
+    let mut register_name: String = "kani_register_loop_contract".to_owned();
+    register_name.push_str(&loop_id);
+    let register_ident = format_ident!("{}", register_name);
+
+    match loop_stmt {
+        Stmt::Expr(ref mut e, _) => match e {
+            Expr::While(ref mut ew) => {
+                let new_cond: Expr = syn::parse(
+                    quote!(
+                        #register_ident(||->bool{#inv_expr}))
+                    .into(),
+                )
+                .unwrap();
+                *(ew.cond) = Expr::Binary(ExprBinary {
+                    attrs: Vec::new(),
+                    left: Box::new(new_cond),
+                    op: BinOp::And(AndAnd::default()),
+                    right: ew.cond.clone(),
+                });
+            }
+            _ => {
+                abort_call_site!("`#[kani::loop_invariant]` is now only supported for while-loops.";
+                    note = "for now, loop contracts is only supported for while-loops.";
+                )
+            }
+        },
+        _ => abort_call_site!("`#[kani::loop_invariant]` is now only supported for while-loops.";
+            note = "for now, loop contracts is only supported for while-loops.";
+        ),
+    }
+    quote!(
+        {
+        // Dummy function used to force the compiler to capture the environment.
+        // We cannot call closures inside constant functions.
+        // This function gets replaced by `kani::internal::call_closure`.
+        #[inline(never)]
+        #[kanitool::fn_marker = "kani_register_loop_contract"]
+        const fn #register_ident<T, F: FnOnce() -> T>(f: F) -> T {
+            unreachable!()
+        }
```

Comment:
```suggestion
        const fn #register_ident<F: FnOnce() -> bool>(_f: F) -> bool {
            unreachable!()
        }
```
BTW, I believe that if you take `f` by reference, you are able to return `true` here and fix #3586.

Reply:
I got the error `error[E0493]: destructor of `F` cannot be evaluated at compile-time`.
I think we can use the same idea of function contracts here that copy the loops when expanding the macro
```
let kani_loop_contract_mode = kani::internal::ORIGINAL
match kani_loop_contract_mode {
   kani::internal::loop_contract_check => #loop_with_loop_contracts_macro
   _ => #original_loop
}
```
And later modify the value of `kani_loop_contract_mode` during transformation.

But I plan to implement this after this PR.

Reply:
What I'm saying is that the error will go away if you change it to a reference though. Something like:
```rust
        const fn #register_ident<F: FnOnce() -> bool>(_f: &F) -> bool {
            true
        }
```

Reply:
Yes, I also added two tests cases showing harness can be correctly proved when loop contracts disabled.